### PR TITLE
WIP: Investigate using `mutagen`

### PR DESCRIPTION
**Pull request by tcharding**
Add mutation testing and mutate the pow module
    
    As an exploration of mutation testing use `mutagen` to mutate the `pow`
    crate. Doing so and running `cargo +nightly mutagen` gives the following
    output (trimmed to show the surviving mutants)
    
    SURVIVED
        bitcoin/src/pow.rs
                19/135(14.07%) mutants survived
            17: replace `>` with `>=` at 407:18-407:19(fn saturating_to_u128)
            18: replace `0` with `1` at 417:21-417:22(fn bits)
            38: replace `||` with `&&` at 437:27-437:29(fn mul_u64)
            39: replace `false` with `true` at 438:33-438:38(fn mul_u64)
            40: replace `false` with `true` at 442:32-442:37(fn mul_u64)
            45: replace `64` with `63` at 446:34-446:36(fn mul_u64)
            49: replace `>` with `>=` at 446:37-446:38(fn mul_u64)
            81: replace `false` with `true` at 507:32-507:37(fn overflowing_add)
            87: replace `>` with `>=` at 532:28-532:29(fn overflowing_sub)
            88: replace `false` with `true` at 545:32-545:37(fn overflowing_mul)
            90: replace `3` with `4` at 547:21-547:22(fn overflowing_mul)
            91: replace `3` with `2` at 547:21-547:22(fn overflowing_mul)
            98: replace `192` with `193` at 553:31-553:34(fn overflowing_mul)
            99: replace `192` with `191` at 553:31-553:34(fn overflowing_mul)
            111: replace `128` with `127` at 643:35-643:38(fn wrapping_shr)
            131: remove `digits 0-9 are valid UTF8` at 677:56-677:83(fn fmt_decimal)
            132: replace `digits 0-9 are valid UTF8` with `-digits 0-9 are valid UTF8` at 677:56-677:83(fn fmt_decimal)
            133: replace `digits 0-9 are valid UTF8` with `digits 0-9 are valid UTF8-` at 677:56-677:83(fn fmt_decimal)
            135: insert `A` at 678:30-678:32(fn fmt_decimal)
    
    I did not explore each error but at least one of them is a false
    positive. This leads me to believe that its a non-trivial amount of work
    to add mutation testing and fix all positives, false or otherwise.
    
    Some other things observed:
    
    - Only builds with nightly so we will need some trickery to be able to run
    tests with stable toolchain still, did not investigate further.
    
    - It wasn't immediately obvious to me exactly which functions were good
    candidates for mutating.

### CODE REVIEW

**Comment by Kixunil** [comment]
Perhaps the strategy should be: annotate a single function, add missing tests, make PR, repeat.

File: bitcoin/Cargo.toml [hunk 1]
```diff
@@ -48,11 +48,12 @@ actual-serde = { package = "serde", version = "1.0.103", default-features = fals
 hashbrown = { version = "0.8", optional = true }
 
 [dev-dependencies]
-serde_json = "1.0.0"
+serde_json = "1.0.91"
 serde_test = "1.0.19"
 serde_derive = "1.0.103"
 secp256k1 = { version = "0.24.0", features = [ "recovery", "rand-std" ] }
 bincode = "1.3.1"
+mutagen = { path = "/home/tobin/build/github.com/llogiq/mutagen/mutagen" }
```
Comment (Kixunil) [inline bitcoin/Cargo.toml:56]: I guess you wanted to use git dep :)

**Comment by apoelstra** [comment]
This looks amazing! Agree with Kix that this will be a *lot* of work and that we probably need to go function by function.

File: bitcoin/Cargo.toml (hunk 1 above)
Comment (tcharding) [reply bitcoin/Cargo.toml:56]: It didn't work at first with git dependency, I just got it working as quickly as possible - thanks though, will fix before doing proper PR.

**Comment by tcharding** [comment]
> Perhaps the strategy should be: annotate a single function, add missing tests, make PR, repeat.

Cool, I'll do the initial infrastructure PR and perhaps one function to get the ball rolling.

**Comment by tcharding** [comment]
> > at least one of them is a false positive.
> 
> How is that even possible? The code changed in a way that it's still valid?

Mutation changed `>` to `>=` and both are logically correct (which is obviously unusual). Not sure how to remedy that since just using `>=` will then allow mutating _back_ to `>` - which again won't break. FTR the function is `saturating_to_u128`.

> > non-trivial amount of work to add mutation testing and fix all positives
> 
> I'm not surprised, the goal of mutation testing is to find missing tests and we probably have a lot of them.
> 
> > Only builds with nightly so we will need some trickery to be able to run
> > tests with stable toolchain still, did not investigate further.
> 
> That's not a big ptoblem, just run nightly

What I meant was having `#[cfg_attr(test, mutate)]` in the code makes tests only runnable with nightly, and that effects work flow - I guess we all have to change our workflow to use `cargo +nightly check` during dev and the `cargo +nightly test` (I don't want to switch between toolchains constantly during dev causing full re-build) 

> > It wasn't immediately obvious to me exactly which functions were good candidates for mutating.
> 
> I would say all non-test, non-trivial functions? For this purpose trivial means one that just passes differently-typed parameters along to some other function without transformation.

Cool, got it.

**Comment by Kixunil** [comment]
> FTR the function is `saturating_to_u128`.

Oh, wow, that's a cool edge case! A bit ugly way to fix it would be to check if high is non-zero and return `u128::max_value()`, else return low.

> What I meant was having `#[cfg_attr(test, mutate)]` in the code makes tests only runnable with nightly

Ah, right, then we should have a cfg for nightly tests or something. `#[fg_attr(all(test, nightly), mutate)]` makes logical sense but it's a bit boilerplate-y.

**Comment by apoelstra** [comment]
I think I'd like to have a mutagen-specific cfg flag rather than calling it `nightly`, but good idea!

**Comment by tcharding** [comment]
So the nightly thing is a bit of a pain, as soon as we add the `mutagen` dependency we can no longer run `cargo test` - even without ever enabling `mutate`.

**Comment by Kixunil** [comment]
Could you try with `optional = true`? And probably use `feature = "mutagen"` on mutagen cfgs.

**Comment by tcharding** [comment]
Oh I didn't think of optional, will try tomorrow. Thanks man.

**Comment by apoelstra** [comment]
`optional` is a good start. I think we can gate it in a stronger way with `[target.'cfg(mutagen)'.dev-dependencies]` or something like that.

**Comment by tcharding** [comment]
`mutagen` is in now, closing this.
